packages/eslint-plugin/src/rules/no-unsafe-enum-assignment.ts

```diff
@@ -0,0 +1,336 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import * as tsutils from 'tsutils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+import { getBaseEnumType, getEnumTypes } from './enum-utils/shared';
+
+const ALLOWED_TYPES_FOR_ANY_ENUM_ARGUMENT =
+  ts.TypeFlags.Unknown | ts.TypeFlags.Number | ts.TypeFlags.String;
+
+export default util.createRule({
+  name: 'no-unsafe-enum-assignment',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow providing non-enum values to enum typed locations',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      operation:
+        'This {{ operator }} may change the enum value to one not present in its enum type.',
+      provided: 'Unsafe non enum type provided to an enum value.',
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const typeChecker = parserServices.program.getTypeChecker();
+
+    /**
+     * Similar to `getEnumTypes`, but returns early as soon as it finds one.
+     */
+    function hasEnumType(type: ts.Type): boolean {
+      return tsutils
+        .unionTypeParts(type)
+        .some(subType => util.isTypeFlagSet(subType, ts.TypeFlags.EnumLiteral));
+    }
+
+    function getTypeFromNode(node: TSESTree.Node): ts.Type {
+      return typeChecker.getTypeAtLocation(
+        parserServices.esTreeNodeToTSNodeMap.get(node),
+      );
+    }
```

Comment:
you call this function for the left and right of each of the assignment cases.
this is going to be pretty expensive across an entire codebase because you're going to be getting a number of "useless" types.

Put another way - there's a number of cases that we can completely ignore purely based on the AST analysis.
For example: `const x = {}` We know we can ignore the `{}` because it's never going to possibly be an enum!

We should use a switch/case here to bail out some cases to improve perf by avoiding type calculation.

```ts
switch (node.type) {
  case AST_NODE_TYPES.ArrayExpression:
  case AST_NODE_TYPES.ArrowFunctionExpression:
  case AST_NODE_TYPES.ClassExpression:
  case AST_NODE_TYPES.ObjectExpression:
  case AST_NODE_TYEPS.FunctionExpression:
  // add any extras we know for sure could never be an enum
   return null;

  default:
   return /* get type */;
}
```

TBH thinking about it you could invert the problem and only check specific cases instead? Like:
- `Identifier`
- `ChainExpression`
- `MemberExpression`
- `CallExpression`
- `ConditionalExpression`
- `Literal`
- `YieldExpression`
- `AwaitExpression`
- `TSAsExpression`
- `TSSatisfiesExpression`
- `AssignmentExpression` (`const x: Enum = y = 1` this code makes me sick)
- `LogicalExpression`
- `BinaryExpression`
- `ImportExpression`
- `TemplateLiteral`
- etc

But there's a lot of cases to check and these cases are probably going to grow over time. So probably more maintainable to include a blocklist instead of an allowlist.

Reply:
A few unhappy counterpoints:

```ts
enum Fruit { Apple }

const fruits: Fruit[] = [0, 1];

function receiveFruits(received: Fruit[]) {}
receiveFruits([0, 1]);

const getFruit: () => fruit = () => 0;
```

I did add `[id.typeAnnotation]` to `VariableDeclarator[init]` so we don't check specifically the case of `const x = {}`. But I don't think there's much else we can do? 😕

Reply:
I guess this is more complicated than I would have liked 😅.

Even class expressions can be fucky:

```ts
interface Foo<T extends Fruit> {
  prop: T;
}

export const clazz: Foo<Fruit> = class WTF {
    static prop: 1 = 1;
}
```

This is going to be a very expensive rule 😢